server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoTileBoundedPredicate.java

```diff
@@ -0,0 +1,118 @@
+
+/**
+ * Filters out tiles using the provided bounds at the provided precision. In order to be efficient it works on the X/Y coordinates of the
+ * geotile scheme.
+ */
+public class GeoTileBoundedPredicate {
+
+    private final boolean crossesDateline;
+    private final long maxTiles;
+    private final int precision, leftX, rightX, minY, maxY;
+
+    public GeoTileBoundedPredicate(int precision, GeoBoundingBox bbox) {
+        this.crossesDateline = bbox.right() < bbox.left();
+        this.precision = precision;
+        if (bbox.bottom() > GeoTileUtils.NORMALIZED_LATITUDE_MASK || bbox.top() < GeoTileUtils.NORMALIZED_NEGATIVE_LATITUDE_MASK) {
+            // this makes validTile() always return false
+            leftX = rightX = minY = maxY = -1;
+            maxTiles = 0;
+        } else {
+            final long tiles = 1L << precision;
+            // compute minX, minY
+            final int minX = GeoTileUtils.getXTile(bbox.left(), tiles);
+            final int minY = GeoTileUtils.getYTile(bbox.top(), tiles);
+            final Rectangle minTile = GeoTileUtils.toBoundingBox(minX, minY, precision);
+            // touching tiles are excluded, they need to share at least one interior point
+            this.leftX = minTile.getMaxX() == bbox.left() ? minX + 1 : minX;
+            this.minY = minTile.getMinY() == bbox.top() ? minY + 1 : minY;
+            // compute maxX, maxY
+            final int maxX = GeoTileUtils.getXTile(bbox.right(), tiles);
+            final int maxY = GeoTileUtils.getYTile(bbox.bottom(), tiles);
+            final Rectangle maxTile = GeoTileUtils.toBoundingBox(maxX, maxY, precision);
+            // touching tiles are excluded, they need to share at least one interior point
+            this.rightX = maxTile.getMinX() == bbox.right() ? maxX : maxX + 1;
+            this.maxY = maxTile.getMaxY() == bbox.bottom() ? maxY : maxY + 1;
+            if (crossesDateline) {
+                this.maxTiles = (tiles + this.rightX - this.leftX) * (this.maxY - this.minY);
+            } else {
+                this.maxTiles = (long) (this.rightX - this.leftX) * (this.maxY - this.minY);
+            }
+        }
+    }
+
+    /** Does the provided bounds crosses the dateline */
+    public boolean crossesDateline() {
+        return crossesDateline;
+    }
+
+    /** The left bound on geotile coordinates */
+    public int leftX() {
+        return leftX;
+    }
+
+    /** The right bound on geotile coordinates */
+    public int rightX() {
+        return rightX;
+    }
+
+    /** The bottom bound on geotile coordinates */
+    public int minY() {
+        return minY;
+    }
+
+    /** The top bound on geotile coordinates */
+    public int maxY() {
+        return maxY;
+    }
+
+    /** Check if the provided tile intersects with the provided bounds. It is assumed that the precision
+     * for the given x/y is the same as the provided in the constructor.*/
+    public boolean validTile(int x, int y) {
```

Comment:
This function is identical to the one that takes `z`, with `precision==z`, so would it not be simpler to have only one function? The cost of the additional `<<` and `+` is probably not high, I would guess.

Reply:
Ok, I removed the other function.